### ISSUE TITLE
Window is not available (SSR next.js)

### ISSUE THREAD

**Issue by favalonx** [firebase/firebase-js-sdk#5547]
### Environment

  * Operating System version: Ubuntu 20.04
  * Browser version: Firefox 92
  * Firebase SDK version: 9.1.0
  * Node version: 16.10
  * Firebase Product: (analytics, performance, remote-config, etc)

#### Steps to reproduce:
Trying to get instances of the products in the Next.js application

#### Relevant Code:

```javascript
import { getRemoteConfig } from 'firebase/remote-config';
import { getPerformance } from 'firebase/performance';
import { firebaseConfig } from '../config/firebase';
import { getAnalytics } from 'firebase/analytics';
import { getFirestore } from 'firebase/firestore';
import { initializeApp } from 'firebase/app';
import { getAuth } from 'firebase/auth';

export const firebaseApp = initializeApp(firebaseConfig);

export const firestore = getFirestore(firebaseApp);
export const auth = getAuth(firebaseApp);

// The three lines below (get* functions to be more precise) produce errors related to SSR like
//   * Window is not available
//   * window is not defined
//   * navigator is not defined
//   * etc.

export const performance = getPerformance(firebaseApp);
// error code: performance/no window

export const analytics = getAnalytics(firebaseApp);
// ReferenceError: window is not defined

export const remoteConfig = getRemoteConfig(firebaseApp);
// error code: remoteconfig/registration-window
```

**Comment by arneesh** [comment]
The window/Node.js env handling is not done internally in the SDK. You would need to handle this on your side via something like by placing a check on window env via an if the condition or a ternary operator.

`typeof window === "undefined"`

**Comment by jamesdaniels** [comment]
RemoteConfig, performance, analytics and others are not supported in Node. You'll need to check `isSupported()` before getting/initializing an instance. 

E,g; `(await isSupported()) ? getRemoteConfig() : undefined`

**Comment by Feiyang1** [comment]
As @jamesdaniels pointed out, you will need to guard the initialization of the browser only packages, and not execute them in Node.js.

**Comment by favalonx** [comment]
I am currently overcoming this issue with the mentioned ways before proposing any of the solutions, thanks to everyone anyway.

What I wanted is to catch the attention of the folks to solve this matter like exactly it was before in SDK v8.

I don't feel it is a good practice to check the window availability and get the instances for every action. Do anyone of you?

**Comment by jamesdaniels** [comment]
> I am currently overcoming this issue with the mentioned ways before proposing any of the solutions, thanks to everyone anyway.
> 
> What I wanted is to catch the attention of the folks to solve this matter like exactly it was before in SDK v8.
> 
> I don't feel it is a good practice to check the window availability and get the instances for every action. Do anyone of you?

I'd suggest to keep the instance around, use `isSupported()` as I mentioned and check for truthiness later on.

```ts
const remoteConfig = (await isSupported()) ? getRemoteConfig() : undefined;
...
if (remoteConfig) {
  // do stuff with remote config
} else {
  // failover behavior (node, etc.)
}
```

**Comment by arneesh** [comment]
> I am currently overcoming this issue with the mentioned ways before proposing any of the solutions, thanks to everyone anyway.
> 
> What I wanted is to catch the attention of the folks to solve this matter like exactly it was before in SDK v8.
> 
> I don't feel it is a good practice to check the window availability and get the instances for every action. Do anyone of you?

Yeah exactly. It was much better in the v8 version. But for the time being this is the state of v9 so we have to use some workaround. I suggested `typeof window === "undefined"` approach because `isSupported()` method is not completely reliable at the moment, it fails on certain instances based on the underlying environment. That's why for the time being I am using the approach I mentioned above. I used this check as a wrapper and inserted all my configs that require browser `env` inside it and I am setting appropriate values accordingly. That being said they are continuously releasing fixes for `isSupported()` method but for the time being, it is not completely reliable. It might work in your case though. But yeah I don't think they will be handling this internally like it was done in the v8 version anytime soon.
